src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs

```diff
@@ -71,74 +76,252 @@ public class SmartDetectorRunner : ISmartDetectorRunner
+            {
+                if (!smartDetectorManifest.SupportedResourceTypes.Contains(alert.ResourceIdentifier.ResourceType))
                 {
-                    this.tracer.TraceError($"Failed running Smart Detector ID {smartDetectorManifest.Id}, Name {smartDetectorManifest.Name}: {e}");
-                    throw new FailedToRunSmartDetectorException($"Calling Smart Detector '{smartDetectorManifest.Name}' failed with exception of type {e.GetType()} and message: {e.Message}", e);
+                    throw new UnidentifiedAlertResourceTypeException(alert.ResourceIdentifier);
                 }
+            }
 
-                // Verify that each alert belongs to one of the types declared in the Smart Detector manifest
-                foreach (Alert alert in alerts)
+            // Trace the number of alerts of each type
+            foreach (var alertType in alerts.GroupBy(x => x.GetType().Name))
+            {
+                this.tracer.TraceInformation($"Got {alertType.Count()} Alerts of type '{alertType.Key}'");
+                this.tracer.ReportMetric("AlertType", alertType.Count(), new Dictionary<string, string>() { { "AlertType", alertType.Key } });
+            }
+
+            // Create results
+            bool detectorSupportsAutomaticResolution = smartDetector is IAutomaticResolutionSmartDetector;
+            List<ContractsAlert> results = new List<ContractsAlert>();
+            foreach (var alert in alerts)
+            {
+                QueryRunInfo queryRunInfo = await this.queryRunInfoProvider.GetQueryRunInfoAsync(new List<ResourceIdentifier>() { alert.ResourceIdentifier }, cancellationToken);
+                ContractsAlert contractsAlert = alert.CreateContractsAlert(
+                    request,
+                    smartDetectorManifest.Name,
+                    queryRunInfo,
+                    this.analysisServicesFactory.UsedLogAnalysisClient,
+                    this.analysisServicesFactory.UsedMetricClient);
+
+                // Handle automatic resolution parameters in the alerts:
+                // If the detector supports automatic resolution - save the predicates for the resolution checks
+                // If the detector doesn't support automatic resolution - drop the automatic resolution parameters (since they are useless) and error trace
+                if (contractsAlert.AutomaticResolutionParameters != null)
                 {
-                    if (!smartDetectorManifest.SupportedResourceTypes.Contains(alert.ResourceIdentifier.ResourceType))
+                    if (detectorSupportsAutomaticResolution)
                     {
-                        throw new UnidentifiedAlertResourceTypeException(alert.ResourceIdentifier);
+                        this.tracer.TraceInformation($"Alert {contractsAlert.Id} has automatic resolution parameters, so saving alert details for later use");
+                        await stateRepository.StoreStateAsync(
```

Comment:
>StoreStateAsync [](start = 46, length = 15)

why the FW should save the state here? why it doesn't come in the API when resolving an alert?

Reply:
talked it over

---
In reply to: [235358901](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/27#discussion_r235358901) [](ancestors = 235358901)